rir/src/interpreter/interp.cpp

```diff
@@ -1232,43 +1232,55 @@ static SEXP seq_int(int n1, int n2) {
     return ans;
 }
 
-#define BINDING_CACHE_SIZE 5
+#define BINDING_CACHE_SIZE 8
+// must be a power of 2 for modulus using & CACHE_MASK to work
+#define CACHE_MASK (BINDING_CACHE_SIZE - 1)
+
 typedef struct {
     SEXP loc;
     Immediate idx;
 } BindingCache;
 
-static RIR_INLINE SEXP cachedGetBindingCell(SEXP env, Immediate idx,
+static RIR_INLINE SEXP cachedGetBindingCell(SEXP env, Immediate poolIdx,
+                                            Immediate cacheIdx,
                                             InterpreterInstance* ctx,
-                                            BindingCache* bindingCache) {
+                                            BindingCache* bindingCache,
+                                            bool smallCache) {
     if (env == R_BaseEnv || env == R_BaseNamespace)
         return NULL;
 
-    Immediate cidx = idx % BINDING_CACHE_SIZE;
-    if (bindingCache[cidx].idx == idx) {
+    Immediate cidx;
+    if (smallCache)
+        cidx = cacheIdx;
+    else
+        cidx = cacheIdx & CACHE_MASK;
+
+    if (bindingCache[cidx].idx == poolIdx) {
```

Comment:
so we changed the cache to contain entries from different environments (to avoid clearing them on change env).

so this is not correct, you would need to check if the environment is correct too. you would need to add the env the cache is for to the bindingCache cell

Reply:
Why? 
In the cache slot we store the unique number determining a pair<symbol,env>.

Reply:
so assume `<"a", e1>` has id 1 and `<"a", e2>` has id 9. they will both end up in slot 1, the check will pass, since both refer to the symbol "a" and you will read the wrong binding.

Reply:
No, because `bindingCache[1].idx = 1` or `bindingCache[1].idx = 9` to distinguish both cases.

Reply:
https://github.com/reactorlabs/rir/pull/471#discussion_r280716465

Reply:
duh, sorry. I did not notice that you changed the index to be the cache index instead of the symbol

Reply:
but the idx field does not contain the cacheIdx, but actually the poolIdx, which is just the symbol, not?

Reply:
also, this check is not needed on small caches. so it should be
```
if (smallCache || bindingCache[cidx].idx == cacheIdx)
```

and then store the cacheIdx into the field everywhere where you update it

Reply:
Ah ok, yeah, the only issue was it should store the cacheIdx yes.
The check is still needed right now because we have to avoid loading of uninit variables